In [2]:
import tensorflow as tf
print(tf.__version__)

1.5.0


In [3]:
%pwd

'/home/howard/ml_tensor'

In [4]:
import numpy as np

In [5]:
np

<module 'numpy' from '/home/howard/ml_tensor/env/lib/python3.6/site-packages/numpy/__init__.py'>

In [6]:
# to make this notebook's output stable across runs
def reset_graph(seed=42):
    tf.reset_default_graph()
    tf.set_random_seed(seed)
    np.random.seed(seed)

## How could you find the partial derivatives of the following function f with regards to x and y? 
###   f=2xy   
###   Let's compute the function at $x=0.2$ and $y=0.3$, and the partial derivatives at that point with regards to $x$ and with regards to $y$:

In [27]:
x = tf.Variable(0.2, name="x")
y = tf.Variable(0.3, name="y")
f=2*x*y

In [28]:
x

<tf.Variable 'x_2:0' shape=() dtype=float32_ref>

In [29]:
y

<tf.Variable 'y_2:0' shape=() dtype=float32_ref>

In [30]:
f

<tf.Tensor 'mul_5:0' shape=() dtype=float32>

In [31]:
learning_rate=0.01
optimizer = tf.train.GradientDescentOptimizer(learning_rate=learning_rate)

In [32]:
optimizer

### Compute gradients of f for the variables in [x,y].
#### grads=optimizer.compute_gradients(f,[x,y])

In [33]:
grads=optimizer.compute_gradients(f,[x,y])
init = tf.global_variables_initializer()
with tf.Session() as sess:
    init.run()
    print(f.eval())
    print(x.eval())
    print(y.eval())
    print(sess.run(grads))

0.120000005
0.2
0.3
[(0.6, 0.2), (0.4, 0.3)]


##  It returns a list of (gradient, variable) pairs

#### $2 \times 0.2\times 0.3=0.12$
#### $f_x=2y=2*0.3=0.6$
####  $f_y=2x=2*0.2=0.4$